apps/plea/models.py

```diff
@@ -639,3 +651,332 @@ class DataValidation(models.Model):
+        ("case_invalid_missing_dateofhearing", "Invalid Case: Missing date of hearing"),
+        ("case_invalid_name_too_long", "Invalid Case: Name too long"),
+        ("case_invalid_offencecode", "Invalid Case: Invalid offence code"),
+        ("case_invalid_courtcode", "Invalid Case: Invalid court code"),
+        ("case_invalid_not_in_whitelist", "Invalid Case: Not in whitelist"),
+        ("case_invalid_duplicate_offence", "Invalid Case: Duplicate offence"),
+        ("case_invalid_duplicate_urn_used", "Invalid Case: Duplicate URN already used (sent)"),
+        ("case_invalid_no_offences", "Invalid Case: No offences"),
+        ("result_invalid_duplicate_urn_used", "Invalid Result: Duplicate URN already used (sent)"),
+    )
+    IGNORED_CASE_FIELDS = [
+        "id",  # Django id is irrelavent
+        "auditevent",  # Case is saved before the auditevent so ignore it
+        "datavalidation",  # DataValidation objects are defunct
+    ]
+    IGNORED_RESULT_FIELDS = ["extra_data", "id"]
+    IGNORED_FORM_FIELDS = ["id"]
+    IGNORED_VALIDATOR_FIELDS = []
+
+    id = models.AutoField(
+        primary_key=True,
+        verbose_name="Audit Event ID")
+
+    case = models.ForeignKey(
+        Case,
+        verbose_name="Related case",
+        help_text="If there was a successful case loaded then it is related here",
+        null=True, blank=True)
+
+    event_type = models.CharField(
+        choices=EVENT_TYPE_CHOICES,
+        verbose_name="event type",
+        help_text="Identified the area of the application that the event happened in",
+        max_length=255)
+
+    event_subtype = models.CharField(
+        choices=EVENT_SUBTYPE_CHOICES,
+        verbose_name="reason for failure",
+        help_text="The specific reason for the event",
+        max_length=255)
+
+    event_trace = models.CharField(
+        max_length=4000,
+        verbose_name="error detail",
+        help_text="This detail about the reason for this event may be useful for developers to debug import issue",
+        blank=True, null=True)
+
+    event_data = HStoreField(
+        verbose_name="Event data",
+        help_text="If there was a failure and data fields were found, they are stored here to debug",
+        null=True, blank=True)
+
+    extra_data_hash = models.CharField(
+        verbose_name="extra data hash",
+        help_text="If the event caused a change to the extra_data then store the hash of it for debugging",
+        max_length=32, default="")
+
+    event_datetime = models.DateTimeField(
+        verbose_name="event date and time",
+        help_text="The time at which the event occurred",
+        auto_now_add=True)
+
+    @property
+    def urn(self):
+        """URN may come from the case or the extra data. Used in the admin."""
+        extra_urn = case_urn = None
+
+        if hasattr(self, "extra_data"):
+            if "urn" in self.extra_data:
+                extra_urn = self.extra_data["urn"]
+
+        if hasattr(self, "case"):
```

Comment:
Can this instance ever not have the `case` attribute? Surely in the least case it would have it but it would be `None`. Could this whole be repalced with `case_urn = self.case and self.case.urn`?